Add S2A Half Connection Counter class

Ryanfsdf opened the pull request:
Added a Counter class for the S2A Half Connection. This will be used in the half connection to keep track of the sequence number which is used as part of the nonce. 

I took the approach of using a byte slice rather than using an integer and converting it to a byte array on the fly whenever `val()` is called. This (using a byte slice) is what is done in ALTS - I'm not sure what the performance differences are but I suspect that this is slightly more efficient than converting an integer on the fly.

Also, I chose to return an error on calling `val` rather than on calling `inc` in case of an overflow, but happy to change that if anyone has other thoughts.

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+
+const counterLen = 12
+
+// counter is a 96-bit, little-endian counter.
```
We need to use a 64-bit sequence number.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+
+import "errors"
+
+const counterLen = 12
```
Since we need to use a 64-bit sequence number, I think this should be 8.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+
+// counter is a 96-bit, little-endian counter.
+type counter struct {
+	value   [counterLen]byte
```
Since we're using a 64-bit sequence number, can we use `uint64` instead? This is similar to the C++ version: 

https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L34

Ryanfsdf replied:
Done, I initially thought that converting on the fly would be inefficient but I think it is actually more efficient. I've changed it to use uint64.

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+// counter is a 96-bit, little-endian counter.
+type counter struct {
+	value   [counterLen]byte
+	invalid bool
```
Can we rename this `hasOverflowed`? I don't think there's any way that a sequence number could be invalid unless it overflows.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+	invalid bool
+}
+
+func newCounter() counter {
```
Please allow the caller to pass in the sequence number as an argument. (In our use cases, the sequence number will always start at zero, but this may change in the future.)

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+	}
+	return c.value[:], nil
+}
+
```
Please add a `reset()` API that sets the sequence number to zero.

Ryanfsdf replied:
Done

matthewstevenson88 commented:
Thanks for the PR Ryan! I've left some initial comments. One general comment: for TLS, the counter is called a sequence number, so please replace `counter` with `sequence` throughout.

Ryanfsdf commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,42 @@
+	return counter{}
+}
+
+// val returns the current value of the counter as a byte slice.
```
I think the `mask()` function should be part of the S2A Half Connection rather than the Counter. The Counter right now provides a method to easily retrieve a byte slice representation of the sequence and also provides overflow checking. I think exposing a `mask` method would break the abstraction that the counter represents so I think it belong to S2A Half Connection rather than the Counter. WDYT?

matthewstevenson88 replied:
SGTM.

Ryanfsdf commented:
Thanks for the review! I've addressed most of the review comments and left a response on one.

Also, I agree that the `counter` should be called `sequence` since in TLS, it is called `sequence`. But I think the class itself should still be called `counter` since I think that more clearly represents what it is doing. However, I think calling the field `sequence` in the actual `S2AHalfConnection` struct may be worthwhile. WDYT?

```
type S2AHalfConnection struct {
	aeadCrypter s2aAeadCrypter
	expander    hkdfExpander
	sequence    counter
}
```

matthewstevenson88 commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,86 @@
+			}
+		}
+
+		// Check that resetting the counter works as expected.
```
Please have separate tests for testing different behaviours.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,48 @@
+	// If the counter is already invalid due to overflow, there is no need to
+	// increase it. We check for the hasOverflowed flag in the call to val().
+	if c.hasOverflowed {
+		return
```
Please throw an error in this case.

Ryanfsdf replied:
I don't think it's necessary to return an error in both `inc()` and `val()`. Should we not return an error in `val()` and instead return an error here?

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,48 @@
+	}
+}
+
+// reset sets the counter value to zero and resets the hasOverflowed flag.
```
s/resets the hasOverflowed flag/sets hasOverflowed to false.

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+// val returns the current value of the counter as a byte slice.
+func (c *counter) val() ([]byte, error) {
```
Per our offline discussion, let's increment the counter as part of this function and remove the `inc()` API.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,48 @@
+	if c.hasOverflowed {
+		return nil, errors.New("invalid counter due to overflow")
+	}
+	buf := make([]byte, counterLen)
```
Per our offline discussion, this should just

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,37 @@
+// counterLen is the byte length of the counter.
+const counterLen = 8
+
+// counter is a 64-bit, little-endian counter.
```
I think we can remove `little-endian` now that there is no more array.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+// counterLen is the byte length of the counter.
+const counterLen = 8
```
I don't think we need this anymore?

Ryanfsdf replied:
Ack. I've removed it for now since it isn't used, but I suspect we'll have to add it back later when implementing the half connection because we need to know how many bytes the counter's uint64 represents.

matthewstevenson88 commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,37 @@
+// getAndIncrement returns the current value of the counter and increments it.
+func (c *counter) getAndIncrement() (uint64, error) {
+	if c.hasOverflowed {
+		return 0, errors.New("invalid counter due to overflow")
```
Could we change this to `counter has overflowed`? We no longer have a concept of an invalid counter.

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,91 @@
+
+type counterTest struct {
+	desc          string
+	counter, want uint64
```
Can we change `want` to `expectedCounter`?

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,91 @@
+			overflow: true,
+		},
+	} {
+		// Test first getAndIncrement call. This should return the same value
```
s/Test/Make

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,91 @@
+			t.Errorf("counter(%v).getAndIncrement() = %v, want %v", test.counter, value, test.counter)
+		}
+
+		// Test second getAndIncrement call. This should verify that the first
```
s/Test/Make

Ryanfsdf replied:
Done

matthewstevenson88 commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,91 @@
+			counter: 123,
+			want:    124,
+		},
+		{
```
Could you add a test vector with `counter: math.MaxUint64 - 1`, `want: math.MaxUint64` and verify that it does not overflow?

Ryanfsdf replied:
Done. I had to add some extra conditions to check for this. I realized that my initial conditions didn't check for the overflow flag in the first call to `getAndIncrement`.

Ryanfsdf commented:
Thanks for the reviews! Addressed all the review comments, PTAL!

matthewstevenson88 commented:
LGTM. Thanks Ryan! Adding Cesar for Go readability.

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func TestCounterInc(t *testing.T) {
+	for _, test := range []counterTest{
```
The two tests don't seem to use the exact same struct. I recommend removing the struct on line 8 and integrate it directly in the tests.

```
for _, tc := range []struct{
   ...
} {

}{

}
```

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func TestCounterInc(t *testing.T) {
+	for _, test := range []counterTest{
```
`tc`? To be consistent with the rest of the code. Unless we used `test` everywhere else.

Ryanfsdf replied:
It seems I've used a mix of `test` and `tc` throughout. Do we have any preference on `tc` vs `test`? We'll need to change it throughout (in a separate PR).

cesarghali replied:
I don't have a preference as long as we're consistent. Let's do the cleanup in a separate PR.

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			// Make first getAndIncrement call. This should return the same value
```
This could already be the case, but please make sure this line is not longer than 80 chars.

cesarghali commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,34 @@
+	hasOverflowed bool
+}
+
+func newCounter(value uint64) counter {
```
In normal cases, will the counter be started from a non-zero value?

Ryanfsdf replied:
I think in normal cases, the counter will start from 0. I'm not sure about any case which starts from a non-zero value right now but I think @matthewstevenson88 had some thoughts on this.

matthewstevenson88 replied:
In TLS 1.3, the counter will always start at zero (unless we have to increment it by a few because of issues with session tickets, but that seems unlikely given the new resumption design). However, in TLS 1.2, the initial sequence number may be nonzero. This is why we should add this plumbing now instead of hardcoding in TLS 1.3 assumptions. WDYT?

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+			c := newCounter(test.counter)
+			value, err := c.getAndIncrement()
+			if err != nil {
+				t.Errorf("counter(%v).getAndIncrement() returned error: %v", test.counter, err)
```
`t.Errorf("counter starting with %v, getAndIncrement() failed, err: %v", test.counter, err)`

cesarghali replied:
The error message implies that `counter(%v)` is a function call.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+				t.Errorf("counter(%v).getAndIncrement() returned error: %v", test.counter, err)
+			}
+			if value != test.counter {
+				t.Errorf("counter(%v).getAndIncrement() = %v, want %v", test.counter, value, test.counter)
```
Same here about replacing `counter(%v)` with what I suggested above.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+				t.Errorf("counter(%v).getAndIncrement() = %v, want %v", test.counter, value, test.counter)
+			}
+			if test.overflow {
+				if !c.hasOverflowed {
```
Can we compare `test.overflow` with `c.hasOverflowed`? This would simplify the `if else` statement.

Ryanfsdf replied:
Good point, I didn't notice that!

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+			}
+			if test.overflow {
+				if !c.hasOverflowed {
+					t.Errorf("counter(%v).getAndIncrement() did not set hasOverflowed flag", test.counter)
```
`t.Errorf("counter starting with %v, c. hasOverflowed = %v, want %v, test.counter, c.hasOverflowed, test.overflow)`

Ryanfsdf replied:
Done

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func TestCounterReset(t *testing.T) {
+	for _, test := range []counterTest{
```
Define a more specific struct inline here.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+func TestCounterReset(t *testing.T) {
```
The goal of `reset` is to set the value to 0 and overflow to false. I suggest we specifically test that without using `getAndIncremenet`. Basically you have access to `counter` so you can construct it the way you want:

`c := &counter{value: test.value, hasOverflowed: test. hasOverflowed}`

The test cases can be:
* non-zero value with false hasOverflowed.
* zero value with false hasOverflowed.
* non-zero value with false hasOverflowed. Not realistic but no harm in testing it.
* zero value with true hasOverflowed.

Ryanfsdf replied:
Done

Ryanfsdf commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+// getAndIncrement returns the current value of the counter and increments it.
+func (c *counter) getAndIncrement() (uint64, error) {
```
I think @matthewstevenson88 had some thoughts on this. The initial PR separated these 2 APIs but we decided to combine them into one to avoid misusing the API.

matthewstevenson88 replied:
I suggested this change because we never want a sequence value to be reused (this would result in the same nonce being used to encrypt/decrypt two different records). To guarantee this, this function changes the sequence number whenever it is accessed. WDYT?

Ryanfsdf commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+// reset sets the counter value to zero and sets hasOverflowed to false.
+func (c *counter) reset() {
```
I believe we'll use this in `KeyUpdate`. Do you have any thoughts on this @matthewstevenson88 ?

matthewstevenson88 replied:
That's correct! It will be called whenever a key update is sent or received.

Ryanfsdf commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,34 @@
+	}
+	val := c.value
+	c.value++
+	if c.value == 0 {
```
Yes correct. Taken from the [TLS 1.3 RFC](https://tools.ietf.org/html/rfc8446#section-5.3):
```
   A 64-bit sequence number is maintained separately for reading and
   writing records.  The appropriate sequence number is incremented by
   one after reading or writing each record.  Each sequence number is
   set to zero at the beginning of a connection and whenever the key is
   changed; the first record transmitted under a particular traffic key
   MUST use sequence number 0.

   Because the size of sequence numbers is 64-bit, they should not wrap.
   If a TLS implementation would need to wrap a sequence number, it MUST
   either rekey (Section 4.6.3) or terminate the connection.
```

Ryanfsdf commented:
I reverted the merging of the API back to `val()` and `inc()` and addressed the review comments, PTAL.

One thing to note is that I also removed the option to specify the value at which the counter starts at. This will make it clearer in the call to `reset` what happens after the value has been reset. Previously, this was a bit ambiguous since it isn't clear whether reset should reset back to the initial value or to 0.

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -34,75 +39,66 @@ func TestCounterInc(t *testing.T) {
-			value, err = c.getAndIncrement()
-			if test.overflow {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counterFromValue(tc.counter)
```
There's no need for a separate function for this. You can do:

```
c := counter{value: tc.counter}
```

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,105 @@
+			val, err := c.val()
+			if tc.overflow {
+				if err == nil {
+					t.Errorf("counter starting with %v, val() expected error, received none", tc.counter)
```
expected overflow error

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,105 @@
+			c := counterFromValue(tc.counter)
+			c.inc()
+			val, err := c.val()
+			if tc.overflow {
```
How about change the code to look like the following:

```
if got, want := (err == nil), !tc.overflow; got != want {
    t..Errorf("counter starting with %v, val()=(err=nil)=%v, want %v", tc.counter, got, want)
}
if err != nil {
    // compare the returned with tc.expectedCounter.
}
```

Also, you might want to consider using the following format in comparisons:

```
if got, want := gotValue, wantValue; got != want {
    t.Errorf("%v: func(%v, %v, ...) = %v, want %v", desc, param1, param2, ..., got, want)
}
```

You're already doing most of the above. I'm only referring to defining `got` and `want`. The benefit is in the error message it's straightforward what's `got` and what's `want`. We can do any necessary cleanup in other files in a separate PR.

Ryanfsdf replied:
Done. I omitted `desc` for your suggestion since that is already given to the Go subtest framework and should print out the description already.

cesarghali replied:
That's totally fine. The error message format I mentioned is the complete one. It can be adjusted as needed. Thanks for the change.

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,105 @@
+			c.inc()
+			val, err := c.val()
+			if tc.overflow {
+				if err == nil {
```
BTW, this case is never triggered. Based on the test cases, the counter will always increment. I suggest adding another field in the test vectors to specify the initial value of the counter's `hasOverflowed` field and add a test case where that is `true`. In this case, you rename the existing `overflow` to `expectedOverflow` or something like that.

Ryanfsdf replied:
I'm not sure what you mean when you say that the case is never triggered. The last test case starts at `MaxUint64`, so wouldn't that trigger the overflow error? I renamed `overflow` to `shouldOverflow` for clarity. Please let me know if I'm missing something!

cesarghali replied:
Ah you're right. I didn't notice line 44 where you increase the counter to get it in the overflow state.

Alternatively and if you want to only test the `val()` API, then you'd start the counter with the state you wanted it at the time `val()` is called and not call `inc()`. Though, the code is fine the way it is.

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,105 @@
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counter{tc.counter, tc.hasOverflowed}
+			// Check that resetting the counter works as expected.
```
No need for this comment since this is implied from the fact that this test exists :)

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/counter_test.go:
```diff
@@ -0,0 +1,90 @@
+			c := counter{value: tc.counter}
+			c.inc()
+			val, err := c.val()
+			if got, want := err == nil, !tc.shouldOverflow; got != want {
```
I know it's not needed but can you change this to `(err == nil)` for clarify? Thanks!

cesarghali commented on security/s2a/internal/crypter/counter.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// val returns the current value of the counter.
+func (c *counter) val() (uint64, error) {
```
Sorry for missing this before, please rename the APIs to `value` and `increment`. It's cleaner that way. Thanks!
